k2/csrc/cuda/ops.h

```diff
@@ -1,71 +1,142 @@
 // k2/csrc/cuda/ops.h
 
-// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey)
+// Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey
+//                                                   Haowen Qiu)
 
 // See ../../LICENSE for clarification regarding multiple authors
 
 #ifndef K2_CSRC_CUDA_OPS_H_
 #define K2_CSRC_CUDA_OPS_H_
 
-#include "k2/csrc/cuda/array.h"
+#include <cassert>
+#include <type_traits>
 
+#include "k2/csrc/cuda/array.h"
 
-// Note, I'm not sure about the name of this file, they are not ops like in TensorFlow, but
-// procedures..
+// Note, I'm not sure about the name of this file, they are not ops like in
+// TensorFlow, but procedures..
+
+// TODO(haowen): this function may be moved to some header files for log (e.g.
+// `log.h`)
+inline cudaError_t CheckCuda(cudaError_t status) {
+#if defined(DEBUG)
+  if (status != cudaSuccess) {
+    fprintf(stderr, "CUDA Runtime Error: %s\n", cudaGetErrorString(status));
+    assert(status == cudaSuccess);
+  }
+#endif
+  return status;
+}
+
+// TODO(haowen): manage/load block config with some classes? then we can get
+// different configuration depending on num_elements and data type.
+constexpr int TILE_DIM = 32;
+constexpr int BLOCK_ROWS = 8;
+
+// TODO(haowen): move the implementations to file `op_inl.h` or
+// `op.cu`(specialized on device and data type)?
+template <typename T>
+__global__ void TransposeKernel(int32_t rows, int32_t cols, const T *input,
+                                T *output) {
+  // TODO(haowen): here we need to handle different type of T to avoid bank
+  // conflicts, the size of cache now is fine for type size with 32bit (e.g.
+  // int32 or float).
+  __shared__ T cache[TILE_DIM][TILE_DIM + 1];
+
+  // input index, in a coalesced manner.
+  int32_t x = threadIdx.x + blockIdx.x * TILE_DIM;
+  int32_t y = threadIdx.y + blockIdx.y * TILE_DIM;
+
+  for (int32_t i = 0; i < TILE_DIM; i += BLOCK_ROWS) {
+    if (x < cols && (y + i) < rows) {
+      cache[threadIdx.y + i][threadIdx.x] = input[(y + i) * cols + x];
```

Comment:
Are these reads coalesced?  IIRC I've seen people loading data, then transposing in local memory, then writing data to disk.

Reply:
yes, it's coalescing as we only change `x` when read data in `input` for a warp.